cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/Gov/TreasuryDonation.hs

```diff
@@ -0,0 +1,149 @@
+      cEra = AnyCardanoEra era
+      fastTestnetOptions = cardanoDefaultTestnetOptions
+        { cardanoEpochLength = 100
+        , cardanoSlotLength = 0.1
+        , cardanoNodeEra = cEra
+        }
+
+  TestnetRuntime
+    { testnetMagic
+    , poolNodes
+    , wallets=wallet0:_
+    , configurationFile
+    }
+    <- cardanoTestnetDefault fastTestnetOptions conf
+
+  PoolNode{poolRuntime} <- H.headM poolNodes
+  poolSprocket1 <- H.noteShow $ nodeSprocket poolRuntime
+  execConfig <- mkExecConfig tempBaseAbsPath poolSprocket1 testnetMagic
+  let socketPath = nodeSocketPath poolRuntime
+
+  epochStateView <- getEpochStateView configurationFile socketPath
+
+  H.note_ $ "Sprocket: " <> show poolSprocket1
+  H.note_ $ "Abs path: " <> tempAbsBasePath'
+  H.note_ $ "Socketpath: " <> unFile socketPath
+  H.note_ $ "Foldblocks config file: " <> unFile configurationFile
+
+  L.Coin currentTreasury <- getTreasuryValue epochStateView
+  H.note_ $ "currentTreasury: " <> show currentTreasury
+  currentTreasury H.=== 0 -- Treasury should initially be 0
+
+  let doOneDonation = doTreasuryDonation sbe execConfig work epochStateView wallet0
+
+  doOneDonation 0 Nothing 500
+  doOneDonation 1 Nothing 500_013
+  doOneDonation 2 Nothing (-497) -- Test donation that should fail because donation is negative
+  doOneDonation 3 (Just 1_234) (-497) -- Test donation that should fail because current treasury value is wrong
+
+doTreasuryDonation :: ()
+  => (HasCallStack, MonadCatch m, MonadTest m, MonadIO m, H.MonadAssertion m)
+  => ShelleyBasedEra era
+  -> H.ExecConfig
+  -> FilePath -- ^ Where temporary files can be stored
+  -> EpochStateView
+  -> PaymentKeyInfo-- ^ The key paying the fee
+  -> Int -- ^ The number of the call, used to create unique temporary file names. Starts at 0.
+  -> Maybe Int -- ^ The current treasury value to use. If unspecified, it will obtained from the node.
+  -> Int -- ^ The amount to donate
+  -> m ()
+doTreasuryDonation sbe execConfig work epochStateView wallet0 idx currentTreasury' treasuryDonation = do
+  currentTreasury <- 
+    case currentTreasury' of
+      Nothing -> do
+        v <- unCoin <$> getTreasuryValue epochStateView
+        H.note_ $ "currentTreasury: " <> show v
+        return v
+      Just x -> pure $ toInteger x
+
+  txBodyFp <- H.note $ work </> "treasury-donation-" <> show idx <> ".body"
+  signedTxFp <- H.note $ work </> "treasury-donation-" <> show idx <> ".signed"
+
+  txIn0 <- findLargestUtxoForPaymentKey epochStateView sbe wallet0
+
+  (exitCode, stdout, stderr) <- execCliAny execConfig
+    [ "conway", "transaction", "build"
+    , "--tx-in", Text.unpack $ renderTxIn txIn0
+    , "--change-address", Text.unpack $ paymentKeyInfoAddr wallet0
+    , "--treasury-donation", show treasuryDonation
+    , "--out-file", txBodyFp
+    ]
+  when (not $ null stdout) (H.note_ stdout)
+  when (not $ null stderr) (H.note_ stderr)
+
+  (exitCode == ExitSuccess) H.=== (currentTreasury >= 0 && treasuryDonation >= 0)
```

Comment:
Quite clever 😄 
But I think we may be lacking logging of the conditions which caused this failure here, I mean the current treasury, donation amount and exitCode.

Reply:
@carbolymer> changed as follows:

```diff
@@ -115,15 +115,22 @@ doTreasuryDonation sbe execConfig work epochStateView wallet0 idx currentTreasur
     , "--treasury-donation", show treasuryDonation
     , "--out-file", txBodyFp
     ]
+  H.note_ $ "exitCode: " <> show exitCode
   unless (null stdout) (H.note_ stdout)
   unless (null stderr) (H.note_ stderr)
 
-  (exitCode == ExitSuccess) H.=== (currentTreasury >= 0 && treasuryDonation >= 0)
-
   case exitCode of
     ExitFailure _ -> do
+      H.assertWith (currentTreasury, treasuryDonation) $ \(ct, td) ->
+        -- If we failed, either the specified treasury was negative,
+        -- or the donation was negative.
+        ct < 0 || td < 0
       return ()
     ExitSuccess -> do
+      H.assertWith (currentTreasury, treasuryDonation) $ \(ct, td) ->
+        -- If we succeeded, both the specified treasury and donation were
+        -- greater or equal to zero.
+        ct >= 0 && td >= 0
```